File: mycollab-caching/src/main/java/com/mycollab/cache/service/impl/InfinispanCacheService.java
package com.mycollab.cache.service.impl;

import com.mycollab.cache.service.CacheService;
import org.apache.commons.collections.CollectionUtils;
import org.infinispan.AdvancedCache;
import org.infinispan.commons.api.BasicCache;
import org.infinispan.context.Flag;
import org.infinispan.manager.EmbeddedCacheManager;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;

import java.util.Set;

/**
 * @author MyCollab Ltd
 * @since 5.1.4
 */
@Service
@Profile({"production"})
public class InfinispanCacheService implements CacheService {
    private static Logger LOG = LoggerFactory.getLogger(InfinispanCacheService.class);

    @Autowired
    private EmbeddedCacheManager instance;

    @Override
    public void putValue(String group, String key, Object value) {
        BasicCache<String, Object> cache = getCache(group);
        cache.put(key, value);
    }

    private BasicCache<String, Object> getCache(String group) {
        BasicCache<String, Object> cache = instance.getCache(group);
        if (cache instanceof AdvancedCache) {
            cache = ((AdvancedCache<String, Object>) cache).withFlags(Flag.IGNORE_RETURN_VALUES);
        }
        return cache;
    }

    @Override
    public Object getValue(String group, String key) {
        BasicCache<String, Object> cache = getCache(group);
        return cache.get(key);
    }

    @Override
    public void removeCacheItem(String id, String prefixKey) {
        BasicCache<String, Object> cache = instance.getCache(id);
        LOG.debug("Remove cache has prefix {} in group {}", prefixKey, id);
        Set<String> keys = cache.keySet();
        if (CollectionUtils.isNotEmpty(keys)) {

            String[] keyArr = keys.toArray(new String[0]);
            for (String aKeyArr : keyArr) {
                if (aKeyArr.startsWith(prefixKey)) {
                    LOG.debug("Remove cache key {}", aKeyArr);
                    cache.remove(aKeyArr);
                }
            }
        }
    }

    @Override
    public void removeCacheItems(String group, Class<?>... classes) {
        for (Class<?> prefKey : classes) {
            removeCacheItem(group, prefKey.getName());
        }
    }
}

File: mycollab-caching/src/main/java/com/mycollab/cache/service/impl/InfinispanCacheService.kt
package com.mycollab.cache.service.impl

import com.mycollab.cache.service.CacheService
import org.infinispan.AdvancedCache
import org.infinispan.commons.api.BasicCache
import org.infinispan.context.Flag
import org.infinispan.manager.EmbeddedCacheManager
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service

/**
 * @author MyCollab Ltd
 * @since 5.1.4
 */
@Service
@Profile("production")
class InfinispanCacheService(private val instance: EmbeddedCacheManager) : CacheService {

    override fun putValue(group: String, key: String, value: Any) {
        val cache = getCache(group)
        cache.put(key, value)
    }

    private fun getCache(group: String): BasicCache<String, Any> {
        var cache: BasicCache<String, Any> = instance.getCache(group)
        if (cache is AdvancedCache<*, *>) {
            cache = (cache as AdvancedCache<String, Any>).withFlags(Flag.IGNORE_RETURN_VALUES)
        }
        return cache
    }

    override fun getValue(group: String, key: String): Any? {
        val cache = getCache(group)
        return cache[key]
    }

    override fun removeCacheItem(group: String, prefixKey: String) {
        val cache = instance.getCache<String, Any>(group)
        LOG.debug("Remove cache has prefix $prefixKey in group $group")
        val keys = cache.keys
        keys.forEach {
            if (it.startsWith(prefixKey)) {
                LOG.debug("Remove cache key $it")
                cache.remove(it)
            }
        }
    }

    override fun removeCacheItems(group: String, vararg classes: Class<*>) {
        classes.forEach { removeCacheItem(group, it.name) }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(InfinispanCacheService::class.java)
    }
}
